### ISSUE TITLE
Turn-off functionality, e.g. for heatpumps

### ISSUE THREAD

**Issue by stevoboy** [rg-engineering/ioBroker.semp#14]
Dear all

I feel there is a turn-off functionality missing, that kicks in, in case the SEMP device does not turn on after the SEMP adapter turns on.

**Use Case SG-Ready heatpump**

1. SEMP Switch in SHM 2.0 is turned off in the beginning
2. SHM plans time-window for the heatpump according the PV surplus projection
3. SHM turns on the SEMP switch through the adapter.
4. The SEMP switch triggers the SG Ready Input in the heatpump

5.
- Case a) Heatpump turns on. SHM/SEMP detects this and leaves the heatpump on according the SHM plan.
- Case b) Heatpump does not turn on because the heat function is deactivtated in summer time, or the warm water / heat level indicates the heatpump does not need to run. In this case the SHM/SEMP detects this and turns off the switch after a pre-defined waiting time. Another device is turned on to optimally use the PV surplus.
- Case c) Heatpump does not turn on within the  pre-defined waiting time. However, it does turn on after the pre-defined waiting time within the time-window defined in SHM/SAE. In this case, other controllable and automatic-restart devices like dishwashers, washing machines are stopped as long as the uncontrollable heatpump is running within its time-windows.
- Case d) Case a) and Case d) happen but the heatpump turns off before the end of its assigned time-window. If this is the case for longer than a pre-defined time (e.g.) 10 minutes, the time-window for the heatpump is closed and another device switched on. This is recognized if the heat-pump power consumption goes below a definable threshold for a defined time.

In my opinion, this functionality is very important as the the is only a limited amount of sun hours in the day, especially in winter times. These should be used to the maximum - especially if a device does not turn on and thereby unnecessarily blocks other devices from starting. I guess this would be also true for wallboxes, where a car might run to 80% charge limit and therefore does not continues charging even though SHM would assign a longer time window than necessary.

**Comment by rg-engineering** [comment]
okay I see the point: if a device requests power from the SHM, but then it won't turn on after the turn on recommendation, the request should also be cleared. I think this can be solved with an option. If a device doesn't follow the recommendation for, say, 10 minutes, then the request is dropped. I think we need the option because a device might turn on much later..

But I actually see two issues: on the one hand, a device does not switch on at all. On the other hand (in your example a wallbox), which turns on, but then turns off after a while. How can one recognize the second case?

**Comment by stevoboy** [comment]
@rg-engineering Sounds good. I put my ideas regarding your last post in case c and d. How does this sound logicwise?

**Comment by stevoboy** [comment]
@rg-engineering One additional request, could you also add a brief description of the data format accepted with the data points for measuring and switch (Objekt-ID Power, Objekt-ID-Schalter, Mindestzeit für neuen Status [Minuten] > better seconds?)? I have assigned a Fritz DECT outlet switch but it appears not to be working maybe because it a true/false but should be 0/1. Knowing the allowed format would help a lot

**Comment by rg-engineering** [comment]
yes of course, but keep in mind the adapter is in alpha-status (not yet in latest repo) so I still fight with functionalities ;-)

**Comment by rg-engineering** [comment]
> Mindestzeit für neuen Status [Minuten] > better seconds

seconds doesn't make sense because SHM requests status once per minute...

**Comment by rg-engineering** [comment]
@stevoboy as a hint for settings have a look here https://github.com/rg-engineering/ioBroker.semp/blob/master/docu/docu_en.md

**Comment by stevoboy** [comment]
> > Mindestzeit für neuen Status [Minuten] > better seconds
> 
> seconds doesn't make sense because SHM requests status once per minute...

Okay, understood :-).

**Comment by stevoboy** [comment]
> yes of course, but keep in mind the adapter is in alpha-status (not yet in latest repo) so I still fight with functionalities ;-)

Okay, but it looks quite mature already. So, I am confident that it goes to repo quickly. Maybe, a status page with the broadcasted values would help for error identification. SAE has something like that /status

**Comment by stevoboy** [comment]
> @stevoboy as a hint for settings have a look here https://github.com/rg-engineering/ioBroker.semp/blob/master/docu/docu_en.md

Looks good. For the switch "javascript.0.semp.Device1_OnOff means that the allowed values are "On" and "Off"? If yes, I need to set-up a custom datapoint with conversion to true/false via Node Red.

**Comment by rg-engineering** [comment]
![grafik](https://user-images.githubusercontent.com/8238668/193645256-42fe68c1-dfe5-4125-b6a6-0a1ba5b93deb.png)

**Comment by stevoboy** [comment]
I got another question, is it already possible to set up multiple time windows?
For my heatpump I would need this to allow morning and afternoon timewindows at least

**Comment by rg-engineering** [comment]
it's on todo-list but not implemented yet...

**Comment by rg-engineering** [comment]
> Maybe, a status page with the broadcasted values would help for error identification. SAE has something like that /status

I added a csv logging to the adapter to see what adapter sent to SHM. Hope this will help debugging...

**Comment by rg-engineering** [comment]
will look like:
![grafik](https://user-images.githubusercontent.com/8238668/196025758-74b1c865-95b7-4061-ab97-57efaccd6531.png)

**Comment by rg-engineering** [comment]
a first (untested) version is available here in github. 
With that version the request is canceled if device does not turn on after a certain time.

**Comment by stevoboy** [comment]
Great! But a stupid question, how do I get this version? It is not 0.0.2, right because installing via url brings me only v.0.0.2:
![image](https://user-images.githubusercontent.com/84937873/196096758-7b8a4e16-4364-4e93-9ecb-26b4f817e087.png)

**Comment by rg-engineering** [comment]
it's still 0.0.2. it's still the same version number...

**Comment by stevoboy** [comment]
Ok, got it. But this Not-turn-on detection does not work with a watts threshold value?

**Comment by stevoboy** [comment]
Unfortunately, the adapter crashes now:

2022-10-18 09:55:04.560 - error: semp.0 (25294) Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch().
2022-10-18 09:55:04.569 - error: semp.0 (25294) unhandled promise rejection: StartCancelRequest is not defined
2022-10-18 09:55:04.572 - error: semp.0 (25294) ReferenceError: StartCancelRequest is not defined
at Device.sendEMRecommendation (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Device.js:613:9)
at Gateway.onSEMPMessage (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Gateway.js:276:13)
2022-10-18 09:55:04.572 - error: semp.0 (25294) StartCancelRequest is not defined
2022-10-18 09:55:05.140 - error: host.iobroker_2022-09-17 Caught by controller[0]: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
2022-10-18 09:55:05.145 - error: host.iobroker_2022-09-17 Caught by controller[0]: ReferenceError: StartCancelRequest is not defined
2022-10-18 09:55:05.145 - error: host.iobroker_2022-09-17 Caught by controller[0]: at Device.sendEMRecommendation (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Device.js:613:9)
2022-10-18 09:55:05.145 - error: host.iobroker_2022-09-17 Caught by controller[0]: at Gateway.onSEMPMessage (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Gateway.js:276:13)
2022-10-18 09:55:05.146 - error: host.iobroker_2022-09-17 instance system.adapter.semp.0 terminated with code 6 (UNCAUGHT_EXCEPTION)
2022-10-18 09:57:04.290 - error: semp.0 (24629) Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch().
2022-10-18 09:57:04.293 - error: semp.0 (24629) unhandled promise rejection: StartCancelRequest is not defined
2022-10-18 09:57:04.296 - error: semp.0 (24629) ReferenceError: StartCancelRequest is not defined
at Device.sendEMRecommendation (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Device.js:613:9)
at Gateway.onSEMPMessage (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Gateway.js:276:13)
2022-10-18 09:57:04.296 - error: semp.0 (24629) StartCancelRequest is not defined
2022-10-18 09:57:04.854 - error: host.iobroker_2022-09-17 Caught by controller[0]: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
2022-10-18 09:57:04.855 - error: host.iobroker_2022-09-17 Caught by controller[0]: ReferenceError: StartCancelRequest is not defined
2022-10-18 09:57:04.855 - error: host.iobroker_2022-09-17 Caught by controller[0]: at Device.sendEMRecommendation (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Device.js:613:9)
2022-10-18 09:57:04.855 - error: host.iobroker_2022-09-17 Caught by controller[0]: at Gateway.onSEMPMessage (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Gateway.js:276:13)
2022-10-18 09:57:04.855 - error: host.iobroker_2022-09-17 instance system.adapter.semp.0 terminated with code 6 (UNCAUGHT_EXCEPTION)
2022-10-18 09:59:04.346 - error: semp.0 (24824) Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch().
2022-10-18 09:59:04.348 - error: semp.0 (24824) unhandled promise rejection: StartCancelRequest is not defined
2022-10-18 09:59:04.351 - error: semp.0 (24824) ReferenceError: StartCancelRequest is not defined
at Device.sendEMRecommendation (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Device.js:613:9)
at Gateway.onSEMPMessage (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Gateway.js:276:13)
2022-10-18 09:59:04.351 - error: semp.0 (24824) StartCancelRequest is not defined
2022-10-18 09:59:04.909 - error: host.iobroker_2022-09-17 Caught by controller[0]: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
2022-10-18 09:59:04.909 - error: host.iobroker_2022-09-17 Caught by controller[1]: ReferenceError: StartCancelRequest is not defined
2022-10-18 09:59:04.909 - error: host.iobroker_2022-09-17 Caught by controller[1]: at Device.sendEMRecommendation (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Device.js:613:9)
2022-10-18 09:59:04.910 - error: host.iobroker_2022-09-17 Caught by controller[1]: at Gateway.onSEMPMessage (/opt/iobroker/node_modules/iobroker.semp/lib/semp/Gateway.js:276:13)
- 2022-10-18 09:59:04.910 - error: host.iobroker_2022-09-17 instance system.adapter.semp.0 terminated with code 6 (UNCAUGHT_EXCEPTION)

**Comment by rg-engineering** [comment]
please try again, I updated it already yesterday (and changed verision number to 0.0.3)

>But this Not-turn-on detection does not work with a watts threshold value?

what do you mean? not-turn-on detection  just checks the status, doesn't matter who sets the status..

**Comment by stevoboy** [comment]
> please try again, I updated it already yesterday (and changed verision number to 0.0.3)
> 
> 
> 
> >But this Not-turn-on detection does not work with a watts threshold value?
> 
> 
> 
> what do you mean? not-turn-on detection  just checks the status, doesn't matter who sets the status..

Running stable without not-turn-on detection since yesterday. Now Checking functionality turned on.

I mean, how do you know the status. If the heatpump starts running you know only by checking the energy meter.

**Comment by rg-engineering** [comment]
> If the heatpump starts running you know only by checking the energy meter.

yes, or you have different datapoints (OID) for status and switch

**Comment by stevoboy** [comment]
> > If the heatpump starts running you know only by checking the energy meter.
> 
> 
> 
> yes, or you have different datapoints (OID) for status and switch

Yes, the OID is the regular case. E.g. dishwasher:
1 6:00 AM load dishwasher, start Programm, SHM turns power plug off until the right window 
2 8:00 AM wife puts coffee cup but leaves the dishwasher door open by accident 
3 10:00 SHM Switches power plug on but dishwasher does not start because the door is open. Energy consumption is below 1 watt 
4 10:10 SEMP terminates window
5 10:11 SHM assigns free time window to another device

Currently this is not working, the dish washer window is still active (window was opened 9:40)

![image](https://user-images.githubusercontent.com/84937873/196898301-f235321c-1652-4ae1-b493-6f770677ff4b.png)

**Comment by stevoboy** [comment]
Ignore the last report of not working please for now... I had some wrong settings. I need to check again tomorrow. I still don't get how the SEMP knows that a device does not start running if not considering a threshold value pointed to the counter.

**Comment by stevoboy** [comment]
Any news regarding how the SEMP detects, if the heatpump is running at all without tracking the power consumption or another boolean switch?

**Comment by rg-engineering** [comment]
From the adapter's point of view, there are only two options: power measurement or separate digital input. The SmartApplianceEnabler is no different. But from an electrical point of view, there are certainly many ways of detecting whether a heat pump is working: separate relay, LED detection, temperature measurements, evaluating the heat pump's status log... But all of this has to take place outside the adapter...
Finally, the SHM and the adapter only need the simple information about the status: on or off.

**Comment by stevoboy** [comment]
> From the adapter's point of view, there are only two options: power measurement or separate digital input. The SmartApplianceEnabler is no different. But from an electrical point of view, there are certainly many ways of detecting whether a heat pump is working: separate relay, LED detection, temperature measurements, evaluating the heat pump's status log... But all of this has to take place outside the adapter... Finally, the SHM and the adapter only need the simple information about the status: on or off.

Yes, agree. But how is this status feedback of the actual running of the heatpump currently integrated in the SEMP. I don't see it anywhere. To me it wouldn't matter if it is a boolean status or number-based status.

**Comment by stevoboy** [comment]
Sorry, my mistakte. I never switched to "From Leistungswert". Do it with the "Vom Leistungswert" what is meant with "Mindestlaufzeit"?

![image](https://user-images.githubusercontent.com/84937873/198571220-35b2cafa-a9dc-4110-bd3c-524f9f3b81a3.png)

**Comment by stevoboy** [comment]
I checked the activity yesterday with the new settings. It looks like the SG-Ready Green Switch was correctly turned on at 10:35 but left on until 16:00 (the end of the time window). However the heatpump did not run the whole day (heating turned off, no warm water needed). 
Shouldn't the Green Switch have turned off a couple of time during the day due to the no-run of the heatpump?

![image](https://user-images.githubusercontent.com/84937873/198863616-93fdce4c-dfab-42b0-b591-44e04c6efe89.png)

**Comment by rg-engineering** [comment]
I implemented that the power request will be aborted if the device doesn't turn on. Can you confirm, that the request was canceled in your case?
The device itself will not switched off yet. That needs to be implemented... Here we have again the case, that SHM will not send a "switch off" command when energy request ends..

**Comment by stevoboy** [comment]
> I implemented that the power request will be aborted if the device doesn't turn on. Can you confirm, that the request was canceled in your case? The device itself will not switched off yet. That needs to be implemented... Here we have again the case, that SHM will not send a "switch off" command when energy request ends..

How can I confirm that the request was cancelled? Which device needs to do that and how?
My heatpump won't do anything actively - I will just turn on the inverter on or off. this happen only once irregularly this morning in the last 24h:

![image](https://user-images.githubusercontent.com/84937873/198867048-e9b40c17-d6d5-4091-8725-5f2b99f1f50f.png)

**Comment by rg-engineering** [comment]
you'll see it in SunnyPortal.  
![grafik](https://user-images.githubusercontent.com/8238668/198867371-e3ae0e72-b214-403e-9f30-d1beeef8bb28.png)

in above sample two devices are planned to run at the same time. When we cancel an request it should disappear in that graph...

**Comment by stevoboy** [comment]
> you'll see it in SunnyPortal.  
> 
> ![grafik](https://user-images.githubusercontent.com/8238668/198867371-e3ae0e72-b214-403e-9f30-d1beeef8bb28.png)
> 
> 
> 
> in above sample two devices are planned to run at the same time. When we cancel an request it should disappear in that graph... 
> 
> 
> 
> 
> 
> 

There is no planning for today:

![image](https://user-images.githubusercontent.com/84937873/198867474-e1af74a3-e26d-4f42-8465-7a6fa894f024.png)

**Comment by stevoboy** [comment]
Again, unfortunately only one True Trigger in the morning and one false in the evening. It appears there is no cancelation after the defined latency time. 

![image](https://user-images.githubusercontent.com/84937873/198897022-485dad3f-955c-4908-9f33-27908527e8b9.png)

The heatpump did not turn on except the one time in the morning not displayed here:
![image](https://user-images.githubusercontent.com/84937873/198897045-ec0ae30f-9273-4e34-88ca-aba678f9bea1.png)

From my point of view the turning on does work, but the switch does not turn of after the predefined waiting time in case on no-run of the heatpump. As we have discussed the SHM does not stop sending true signals itself in case of a no-run but keeps sending true signals throughout the day. 

In my point of view purpose of the SEMP is to send a cancelation of time window to the SHM after the pre-defined time of no-run. The next chance for a run would be with the beginning of the next time window. So, I would expect to set-up hourly timewindows, that are closed with SEMP after the pre-defined waiting time. The heatpump would then get the next chance at the beginning of every hour.

**Comment by rg-engineering** [comment]
> From my point of view the turning on does work, but the switch does not turn of after the predefined waiting time in case on no-run of the heatpump

yes, as I wrote: the implementation to turn off the device is still missing. I just cancel the energy request. That means I do not send energy requests anymore to SHM. So the SHM should then stop to send "turn on" commands. In next version the adapter will switch off the device too, but I need to overwork my source code first... 

< In my point of view purpose of the SEMP is to send a cancelation of time window to the SHM after the pre-defined time of no-run

this is exactly what I've implemented...

> So, I would expect to set-up hourly timewindows

this is up to you with next version. I'm implementing the support of more then one time window per day. The feature will come with next version...

**Comment by stevoboy** [comment]
Alright sounds great. Just to clarify what I have in my mind with a simple example:

Time window 1: 09:00 - 10:00
Time window 2: 10:00 - 11:00
Time window 3: 11:00 - 12:00

Scenario 1: SHM turns switch on at 9:00. SEMP detects this. Heatpump starts running for three hours. SEMP does nothing, all three time windows stay open.

Scenario 2: SHM turns switch on at 9:00. SEMP detects this. Heatpump does not start running. After defined latency time, SEMP sends close window command to SHM. The 9:00 - 10:00 time window closes early. SHM could plan other devices like a wallbox in the meantime.

Scenario 3: SHM turns switch on at 9:00. SEMP detects this. Heatpump starts running but stops at 9:30. SEMP detects this. After defined latency time, SEMP send close window command to SHM. The rest of the time window can be used for other devices.

And so, basically the detection of the actual running is by the boolean switch or watts counter. The user definds the latency time until which Semp waits counting from the beginning of the time window to shut down the time window.

**Comment by stevoboy** [comment]
Btw. Today I have a planning by SHM for the heatpump 

![image](https://user-images.githubusercontent.com/84937873/198947048-ba40f57b-a042-43b0-a911-c757b6866de0.png)

**Comment by rg-engineering** [comment]
So far I have understood the use case. In general, you can do this when I have the next version ready that will support multiple energy request windows.
But: You have to consider that the SHM makes its own plans. So if you want the SHM to turn on a device at a specific time, the MinRunTime and the MaxRunTime must exactly match the length of the time window.
In any case, the check as to whether a device switches on only begins when the SHM switches the device on. It makes no sense to start the test at the beginning of the configured time window, since the SHM may plan to switch on the device later. I can't get any information about the plans of the SHM in the adapter.

By the way, you can read the details regarding the planning and time frame of the SHM at [SEMP Protocol](https://github.com/rg-engineering/ioBroker.semp/blob/master/docu/SMA/SEMP-11ZE3315-Specification-1.0.6.pdf) on page 43

and again: with your current version a time frame should disappear from portal graph, if device does not turn on. Would be good to know whether that works already...

**Comment by stevoboy** [comment]
Alright. Sounds great. I check the point regarding the SEMP protocol and timeframe disappearing and report back here.

**Comment by stevoboy** [comment]
I monitored the behaviour in the Sunny Portal today. It does not work unfortunately. The planning window changes once but in total it stays the same. Also the SEMP switch does not turn-off the whole day despite the heatpump not running and other devices are turned on. There was no real surplus energy today.

![image](https://user-images.githubusercontent.com/84937873/199603440-e878c4b3-0006-4e5e-ac03-2c3dd38b2ca9.png)

![image](https://user-images.githubusercontent.com/84937873/199603480-0ae653e4-c2dc-40da-8403-2187a49d6709.png)

![image](https://user-images.githubusercontent.com/84937873/199604205-2c397bec-1acd-4a2d-b864-7af3156ef216.png)

![image](https://user-images.githubusercontent.com/84937873/199604231-10e96287-1efe-405b-8cca-99d8be5ebcfe.png)

![image](https://user-images.githubusercontent.com/84937873/199604282-1f8da1b1-faf2-4f92-9f41-135e21b0a05d.png)

![image](https://user-images.githubusercontent.com/84937873/199604310-a551c274-56a1-45e9-9551-902a8c5bb522.png)

![image](https://user-images.githubusercontent.com/84937873/199604349-a0dcf2be-ee5a-48aa-bc11-651a90e50260.png)

**Comment by rg-engineering** [comment]
okay, will check it again now. I'm almost finished with multiple time frames per day day now, so I can use my latest version for further tests..

**Comment by stevoboy** [comment]
Sounds great. Let me know when I can test the new version.

I checked again the IOB logs of the SEMP switch today - same behaviour as yesterday... true setting early in the morning and no false until end of day.

![image](https://user-images.githubusercontent.com/84937873/199818786-cfffb058-cd0b-41fa-8e7a-d4f5c90c3e0a.png)

**Comment by rg-engineering** [comment]
@stevoboy Können wir das hier schliessen, oder hast du neue Erkenntnisse?

**Comment by stevoboy** [comment]
Ja, kann geschlossen werden.
Ich habe das Projekt aus zeitgründen nicht weiter verfolgt und bin schlussendlich auf eine Fritzsteckdose gegangen.
